### DISCUSSION TITLE
Fetch remote & fast forward keybindings

### DISCUSSION THREAD

**Discussion by MatejBransky** [jesseduffield/lazygit#3553] [Q&A] [answered]
Hi, thank you for the best git UI!

I've updated to the latest version and before that I used `fetch remote` a lot. But now the same keybinding is applied to the `fast forward`. 

I've noticed that `fetchRemote` and `fastForward` are mapped in the same context (`branches`) and that they have [the same keybinding](https://github.com/jesseduffield/lazygit/blob/master/docs/Config.md?plain=1#L252). 

I've tried to change the `fastForward` to a different keybinding but the `fetchRemote` is still not callable. 

Is there something what I've missed or is this a bug?

Thanks for any response. 🙏

**Comment by stefanhaller** [comment]
Which older version have you updated from? I don't see that "f" has ever been used for fetching in the branches panel. The fact that the `fetchRemote` keybinding is listed under "branches" can be seen as a bug in the config; the keybinding is only used in the remotes tab, for fetching just the selected remote. (The remotes tab doesn't have its own keybinding section though, and it appears in the branches panel, so it makes _some_ sense to list it there.)

There is a `Fetch` command in the files panel (also bound to "f", maybe a bit confusingly), is that what you have been using?

**Reply by MatejBransky** [reply]
That's a problem. I don't remember the version before. I just know that it's started to work differently from [this release](https://github.com/jesseduffield/lazygit/releases/tag/v0.41.0).

Nevertheless I've created custom command which does the same thing but with a little bit more invisible status (visible only in the command pane).

```yaml
customCommands:
  - key: <c-f>
    description: Fetch branch
    context: localBranches
    command: git fetch origin {{.SelectedLocalBranch.Name}}:{{.SelectedLocalBranch.Name}}
```

**Reply by stefanhaller** [reply, marked as answer]
Uhm, that command does exactly what the `fastForward` command does, which is bound to "f" in the branches panel.

**Reply by MatejBransky** [reply]
You're right. 

So now I'm confused. 🤔 Maybe it's my misunderstanding of the git fetch. I thought that it just fetches the latest version of the branch selected in the branch panel. I've thought that it changed because it started to show some unstaged changes in the currently checkouted branch and I didn't know why it interferes into my checkouted branch. 🤷 

Sorry for taking your time and thank you for your help. 🙏

**Reply by stefanhaller** [reply]
Yes, all of this can be very confusing. The command-line API design of git is confusing: a `git fetch` on its own fetches all remote branches, but leaves the local branches alone (and that's typing "f" in the Files panel in lazygit). `git fetch origin x:x` does almost the opposite, it only fetches the selected branch (both the remote branch, and updates the local branch), but leaves all other remote branches alone. I find this highly confusing, but it's useful. :-)

On top of that, typing "f" in the branches panel in lazygit does different things depending on whether you do it on the checked-out branch or another one. For the checked-out branch, it actually calls `git pull`. And this could explain why it affected your working copy (for example if the .gitignore file has changed upstream).
